### PR TITLE
Reorganize cuSpatial headers

### PR DESCRIPTION
**Pull request by harrism**
## Description
Fixes https://github.com/rapidsai/cuspatial/issues/1080 based on the discussion in #1087

-  Consolidate each major set of features within common headers, e.g. `distance.hpp/.cuh`, `intersection.hpp/.cuh`.
-  Consolidate point-in-polygon APIs / headers and DRY up implementation
- Rename some headers/APIS that were inconsistent across the header-only and column-based APIs, e.g. `points_in_spatial_window` vs. `points_in_range`.
- Update Cython where needed to account for above changes.

This PR does NOT yet reorganize source code.

Update: Compilation Time Results
---------------------------------------------

Compilation time is not changed appreciably by this PR (<0.5% faster).

This test was performed by disabling `sccache`, erasing everything in the build directory, configuring cmake (including building tests and benchmarks), and then running `time ninja`. My machine has a "AMD Ryzen 7 3700X 8-Core Processor", and I'm using Ninja's default parallelism.

#### Disabling `sccache` when using the cuSpatial devcontainer

```
(rapids) coder ➜ ~/cuspatial/ $ cd cpp/build/release
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ rm -rf ./*
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ unset RUSTC_WRAPPER
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ unset CMAKE_C_COMPILER_LAUNCHER
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ unset CMAKE_CXX_COMPILER_LAUNCHER
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ unset CMAKE_CUDA_COMPILER_LAUNCHER
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ cmake ~/cuspatial/cpp -GNinja -DBUILD_TESTS=ON -DBUILD_BENCHMARKS=ON
(rapids) coder ➜ ~/cuspatial/cpp/build/release $ time ninja
```

### This branch: 

```
[201/201] Linking CXX executable gtests/LINESTRING_INTERSECTION_TEST_EXP

real    10m42.845s
user    122m28.077s
sys     6m7.935s
```

### `branch-23.06`:

```
[202/202] Linking CXX executable gtests/LINESTRING_INTERSECTION_TEST_EXP

real    10m45.573s
user    122m52.896s
sys     6m9.357s
```

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Review by isVoid — APPROVED** [review]
Only small comments below, I think we should merge this asap so that you don't have to deal with conflicts from other C++ changes.

1 more comment on cython pxd bindings ("_lib/cpp/") - their file structure should exactly match hpp files. But I can create a new issue to track that.

**Comment by isVoid** [comment]
Can you also provide compile time of this PR against branch-23.06?

File: cpp/include/cuspatial/detail/bounding_boxes.cuh [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  inline CUSPATIAL_HOST_DEVICE box<T> operator()(box<T> const& a, box<T> const& b)
+  {
+    return {box_min(box_min(a.v1, a.v2), b.v1), box_max(box_max(a.v1, a.v2), b.v2)};
+  }
+};
+
+}  // namespace detail
+
+template <typename IdInputIt, typename PointInputIt, typename BoundingBoxOutputIt, typename T>
+BoundingBoxOutputIt point_bounding_boxes(IdInputIt ids_first,
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cuspatial/detail/bounding_boxes.cuh:62]: I think combining implementations only makes sense if it reduces complexity and repetition. If it causes large files that are harder to navigate then I wouldn't do it.  I've seen both situations in cuSpatial, so I think we can handle it on a case-by-case basis.

**Comment by harrism** [comment]
> Can you also provide compile time of this PR against branch-23.06?

Added to description.

**Review by trxcllnt — APPROVED** [review]

**Review by thomcom — APPROVED** [review]

**Pull request merged**